style_guides/pl_sql_general_guidance.md

```diff
@@ -0,0 +1,175 @@
+...
+```
+
+__BAD__:
+
+```sql
+SELECT
+    projects.name project_name,
+    COUNT(backings.id) backings_count
+...
+```
+
+Order (`ASC`, `DESC`) should always be explicit. All window functions should be aliased.
+
+---
+
+## Table Joins
+
+__USE ANSI JOIN SYNTAX!__
+
+Explicitly use `INNER JOIN` not just `JOIN`, making multiple lines of `INNER JOIN`s easier to scan:
+
+__GOOD__:
+
+```sql
+SELECT
+    projects.name      AS project_name,
+    COUNT(backings.id) AS backings_count
+FROM 
+    ksr.projects AS projects
+    INNER JOIN ksr.backings AS backings
+    ON ...
+    LEFT OUTER JOIN ksr.backer_rewards AS backer_rewards 
+    ON ...
+    LEFT OUTER JOIN ...
+```
+
+__BAD__:
+
+```sql
+SELECT
+    projects.name      AS project_name,
+    COUNT(backings.id) AS backings_count
+FROM 
+    ksr.projects AS projects
+    JOIN ksr.backings AS backings 
+    ON ...
+    LEFT OUTER JOIN ksr.backer_rewards AS backer_rewards
+    ON ...
+    LEFT OUTER JOIN ...
+```
+
+---
+
+## Transaction Control COMMIT and ROLLBACK 
+
+A function or procedure should take a parameter to specify this behavior. This means it can be controlled by the calling code and makes it more flexible.
+
+In your package spec specify constants for commit and rollback:
+
+```sql
+g_COMMIT   CONSTANT NUMBER := 1;
+g_ROLLBACK CONSTANT NUMBER := -1;
+```
+
+Pass this value as a parameter into functions and procedures:
+
+```sql
+PROCEDURE get_transactions(i_commit IN NUMBER)
+```
+
+It can then be used like this:
+
+```sql
+IF p_commit = g_COMMIT THEN
+    DSP.line('package.get_transactions', 'Commit');
+    COMMIT;
+ELSIF p_commit = g_ROLLBACK THEN
+    DSP.line('package.get_transactions', 'Rollback');
+    ROLLBACK;
+END IF;
+```
+
+## Table Sequences
+
+From Oracle Database 12c use database identity columns to define table sequence columns:
+
+```sql
+ CREATE TABLE new_table (
+    primary_key_id NUMBER GENERATED BY DEFAULT ON NULL AS IDENTITY,
+    ...
+);
+```
+
+The value in column primary_key_id will be autmoatically generated as an incrementing value. For more information on this visit: [docs.oracle.com](https://docs.oracle.com/en/database/other-databases/nosql-database/22.1/sqlreferencefornosql/identity-column.html)
+
+
+## Single Responsibility Principle
+
+When writing code always consider the [Single responsibility principle](https://en.wikipedia.org/wiki/Single_responsibility_principle). Each block of code should do one thing, and one thing well.
+
+
+## Conditional Statements
+
+Unless your use case absolutely requires a `DECODE`, always use a `CASE` statement for conditional processing. 
+
+In all but the simplest scenarios a `CASE` satatement is more readable ( and therefore maintainable ) and handles a larger variety of conditions.
+
+For more information on using `CASE` statements, please see [The Difference Between DECODE and CASE](https://www.oratable.com/decode-case-differences/)
+
+__GOOD__:
+
+```sql
+dept_desc := CASE deptno

```

Comment:
Maybe move the CASE to the line below in line with the END. This looks awful :)

Reply:
I actually prefer this layout when using a case statement like this. guess it will be whatever the auto formatter defaults too.

Reply:
No problem, I will run all the code on this page through the auto formatter and we'll leave it at that